prometheus-cpp: 0.9.0 -> 1.0.0

cole-h opened the pull request:
###### Description of changes

I want to use the Exposer::GetListeningPorts function, but it was just
released in 1.0.0.

This required essentially redoing the civetweb package to use cmake instead of make (see the commit message for information on the exact error).

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

~~This breaks `proxysql` because it wants and patches 0.9.0 :')~~ We use their vendored prometheus-cpp (for) now.

cole-h committed:
civetweb: use cmake

prometheus-cpp 1.0.0 depends on civetweb's cmake files being available:
    CMake Error at pull/CMakeLists.txt:11 (find_package):
      Could not find a package configuration file provided by "civetweb" with any
      of the following names:

        civetwebConfig.cmake
        civetweb-config.cmake

      Add the installation prefix of "civetweb" to CMAKE_PREFIX_PATH or set
      "civetweb_DIR" to a directory containing one of the above files.  If
      "civetweb" provides a separate development package or SDK, be sure it has
      been installed.

Cmake also properly handles setting the prefix, libdir, and includedir
paths, so that patch was dropped.

cole-h committed:
prometheus-cpp: 0.9.0 -> 1.0.0

I want to use the Exposer::GetListeningPorts function, but it was just
released in 1.0.0.

cole-h commented:
cc @ajs124 -- since you originally packaged `proxysql`, do you have any ideas on how difficult it would be to get upstream to use `prometheus-cpp` 1.0.0? Otherwise, it may be necessary to keep 0.9.0 packaged separately as `prometheus-cpp_0_9`.

ajs124 commented:
We could just "re-vendor" it. proxysql ships basically all of its dependencies as tarballs in their git repo, I've just opted to use most of them from nixpkgs instead. See [this patch](https://github.com/NixOS/nixpkgs/blob/master/pkgs/servers/sql/proxysql/makefiles.patch#L144) and some part of the derivation where their prometheus-cpp is replaced with ours.

Besides that, they've been mostly responsive, when I've interacted with them, so it's worth a try, just opening an issue.
Or maybe somebody could open a PR, if it's not too big of an API change.

cole-h commented:
I'll file an issue against their repo.

By "re-vendor", do you mean just using the sources they have tarballed in-tree, or something else?

ajs124 commented:
That's what I meant, yes.

cole-h commented:
Ready for review -- I've re-vendored `prometheus-cpp` while we wait for upstream to decide on whether or not they'll upgrade to prometheus-cpp 1.0.0. At that point, we can probably revert `proxysql: use vendored prometheus-cpp`.

---

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>civetweb</li>
    <li>minetestserver</li>
    <li>prometheus-cpp</li>
    <li>proxysql</li>
  </ul>
</details>

cole-h commented on pkgs/servers/sql/proxysql/default.nix:
```diff
@@ -123,8 +121,9 @@ stdenv.mkDerivation rec {
-    pushd prometheus-cpp/prometheus-cpp/3rdparty
-    replace_dep . "${civetweb.src}" civetweb
+    pushd prometheus-cpp
+    tar xf v0.9.0.tar.gz
```
(Diffs on in-tree diffs are always a mess lol)

Oh, it seems you're right. I think we have 2 paths forward:

1) we could also use their civetweb vendor and just remove this entire prometheus-cpp section
2) or we could undo my diff on the `makefiles.patch` and keep this manual un`tar`ing of the archive

Do you have a preference?

ajs124 replied:
I think I'd prefer option 2, assuming it works

cole-h replied:
Done in [`90a50ca` (#163695)](https://github.com/NixOS/nixpkgs/pull/163695/commits/90a50caa4e439e4c9fe9e6cb098d0556abdc7720). It seems to work (build succeeds at the very least).

cole-h committed:
proxysql: use vendored prometheus-cpp

They specifically want v0.9.0 and already have it vendored. However,
we continue to use the civetweb from Nixpkgs. It may be necessary at
some point in the future to just allow them to use their vendored
sources for this altogether, but it works fine for now.

SuperSandro2000 commented on pkgs/development/libraries/civetweb/default.nix:
```diff
@@ -10,36 +12,31 @@ stdenv.mkDerivation rec {
     repo = pname;
     rev = "v${version}";
-    sha256 = "sha256-Qh6BGPk7a01YzCeX42+Og9M+fjXRs7kzNUCyT4mYab4=";
+    sha256 = "Qh6BGPk7a01YzCeX42+Og9M+fjXRs7kzNUCyT4mYab4=";
```
```suggestion
    sha256 = "sha256-Qh6BGPk7a01YzCeX42+Og9M+fjXRs7kzNUCyT4mYab4=";
```
or
```suggestion
    hash = "sha256-Qh6BGPk7a01YzCeX42+Og9M+fjXRs7kzNUCyT4mYab4=";
```

ajs124 reviewed (approved):
`proxysql` still builds and starts, which is good enough for me.
we don't use the prometheus exporter anyway, tbh.
